Feature/ddm 96

minseonglove opened the pull request:
### 개요
로그인 버튼을 누르면 api를 호출한 뒤 응답으로 "success"를 받았다면 메인 화면으로 넘어갑니다.

### 변경사항
로그인 성공 시 넘어갈 메인 화면 추가 (MainFragment.kt, fragment_main.xml)
상수 클래스 추가 (Constants.kt)
http 동작들을 정의한 인터페이스 추가 (RetrofitService.kt)
Retrofit.Builder 클래스를 설정해줄 object 클래스 추가 (RetrofitInstance.kt)
생명주기에 관계없이 비동기 api를 처리할 뷰모델 클래스 추가 (MainViewModal.kt)
뷰모델을 생성하기 위한 팩토리 클래스 추가 (MainViewModalFactory.kt)


### 관련 지라 및 위키 링크
[DDM-96](https://fake-developers.atlassian.net/jira/software/projects/DDM/boards/10)

### 리뷰어에게 하고 싶은 말
LoginFragment.kt에서 로그인에 대한 응답 메세지 검증을 쌩 문자열 "success"를 비교해서 하고 있는데
현재 호출하고 있는게 테스트 api이고 언제든 다른 api로 교체된다는 점을 감안하여
따로 상수나 enum으로 처리하지 않은 점 이해 바랍니다.

minseonglove committed:
FEAT : DDM-96 AOS 로그인 화면 만들기

retrofit2 라이브러리를 이용해 서버에 login api를 호출한 뒤 "success"를 리턴 받으면 메인 화면으로 넘어갑니다.

minseonglove committed:
FEAT : DDM-96 AOS 로그인 화면 만들기

minseonglove committed:
FEAT : DDM-96 AOS 로그인 화면 만들기

konempty commented on app/build.gradle:
```diff
@@ -35,12 +35,21 @@ android {
+    // Converter (JSON 타입 결과를 객체로 매핑)
+    implementation 'com.squareup.retrofit2:converter-gson:2.9.0'
+    // Converter (결과가 그냥 문자열 일 때)
+    implementation 'com.squareup.retrofit2:converter-scalars:2.6.0'
```
얘만 버전이 다른가요? 확인부탁드릴게요!

minseonglove replied:
> 얘만 버전이 다른가요? 확인부탁드릴게요!

2.9.0이 있었네요 시정하겠습니다

konempty commented on app/src/main/java/com/fakedevelopers/ddangddangmarket/MainActivity.kt:
```diff
@@ -26,6 +27,10 @@ class MainActivity : AppCompatActivity() {
                 transition.addToBackStack(null)
             }
+            FragmentType.MAIN -> {
+                transition.replace(R.id.mainContainer, mainFragment)
```
뭔가 코드가 똑같은게 3개가 있는 느낌인데 이걸 쫌더 예쁘고 짧게 가능할것 같네요

minseonglove replied:
> 뭔가 코드가 똑같은게 3개가 있는 느낌인데 이걸 쫌더 예쁘고 짧게 가능할것 같네요

두뇌 풀가동해서 줄여보겠읍니다.

konempty replied:
지금 떠오르는건 Map<FragmentType, Fragment> 형태의 맵을 쓰면 딱 두절? 이면 될듯 하네요

minseonglove replied:
> 지금 떠오르는건 Map<FragmentType, Fragment> 형태의 맵을 쓰면 딱 두절? 이면 될듯 하네요

앗 enum에 get함수를 줘서 enum 자체가 프래그먼트를 반환하도록 바꿨읍니다.

konempty commented on app/src/main/java/com/fakedevelopers/ddangddangmarket/api/instance/RetrofitInstance.kt:
```diff
@@ -0,0 +1,24 @@
+    private val retrofit by lazy {
+        Retrofit.Builder()
+            .baseUrl(BASE_URL)
+            .addConverterFactory(ScalarsConverterFactory.create())
```
오..? 그냥 스트링은 이게 있어야 인식이 되나요?

minseonglove replied:
Json은 GsonConverterFactory
그냥 스트링은 저게 있어야 인식됩니다.
gson으로 받으니까 팅겨 버리드라구요.

konempty replied:
오호라 처음알았네요 그건 ㅋㅋㅋㅋㅋ

konempty commented on app/src/main/java/com/fakedevelopers/ddangddangmarket/api/service/RetrofitService.kt:
```diff
@@ -0,0 +1,12 @@
+interface RetrofitService {
+    @FormUrlEncoded
+    @POST("user/login")
+    suspend fun loginRequest(@Field("email") email: String, @Field("passwd") passwd: String): Response<String>
```
변수명과 필드명이 같으면 굳이 안써도 될것 같은데 확인부탁합니다 ㅋㅋㅋㅋ

minseonglove replied:
> 변수명과 필드명이 같으면 굳이 안써도 될것 같은데 확인부탁합니다 ㅋㅋㅋㅋ

필드명과 변수명이 같으면 안써도 됩니다.
하지만 저렇게 명시를 해두는게 보기도 이쁘고 실수도 방지할 것 같아서 냅뒀습니다.

konempty commented on app/src/main/java/com/fakedevelopers/ddangddangmarket/ui/login/LoginFragment.kt:
```diff
@@ -1,23 +1,62 @@
+            // 결과 처리
+            viewModel.loginResponse.observe(viewLifecycleOwner) {
+                if(it.isSuccessful){
+                    Log.d("Login", it.body().toString())
```
로그는 취향 차이긴한데 로그 라이브러리도 찾아보시면 좋을듯 합니다

minseonglove replied:
> 로그는 취향 차이긴한데 로그 라이브러리도 찾아보시면 좋을듯 합니다

확실히 기본 로그는 멋없긴하네요 다른 로그로 바꾸겠읍니다

konempty commented on app/src/main/res/values/strings.xml:
```diff
@@ -10,4 +10,6 @@
     <string name="login_password">비밀번호</string>
     <string name="login_signin">시작하기</string>
+    <!-- Strings related to main -->
+    <string name="main_loginSuccessText">로그인 성공!</string>
```
저번에도 말했지만 문자열은 다른 프론트 개발자들과 협의후 추가하시는걸 추천드립니다 :)

minseonglove replied:
> 지금 봤는데 PR제목도 조금 신경써 주시고 커밋메시지에서 한줄은 최대 72Byte로 맞춰줘야 할거예요 ㅋㅋㅋ 그부분도 다음부턴 신경써주세여. pr내용 각줄에 - 를 붙여주면 보기 좋을것 같네요
> 
> * 이렇게요
> * ㅋㅋㅋㅋㅋ

다음엔 더 예쁜 PR로 찾아 뵙겠읍니다.

konempty reviewed (changes requested):
수고하셨습니다! 리뷰 확인후 말씀해주세요

konempty commented:
지금 봤는데 PR제목도 조금 신경써 주시고 커밋메시지에서 한줄은 최대 72Byte로 맞춰줘야 할거예요 ㅋㅋㅋ 그부분도 다음부턴 신경써주세여. pr내용 각줄에 - 를 붙여주면 보기 좋을것 같네요
- 이렇게요
- ㅋㅋㅋㅋㅋ

minseonglove committed:
REFACTOR : konempty PR 리뷰 반영

> * Logger 라이브러리 설치
> * converter-scalars 라이브러리 버전업
> * setFragment 메소드 최적화

konempty commented on app/build.gradle:
```diff
@@ -35,12 +35,23 @@ android {
+    // Converter (결과가 그냥 문자열 일 때)
+    implementation 'com.squareup.retrofit2:converter-scalars:2.9.0'
+    // okhttp3
+    implementation 'com.squareup.okhttp3:logging-interceptor:5.0.0-alpha.6'
```
??? 알파버전은 쓰는건 흠... 이건뭐 취향문제라서 맘대로하세요ㅋㅋㅋㅋㅋ

konempty commented on app/src/main/java/com/fakedevelopers/ddangddangmarket/FragmentType.kt:
```diff
@@ -1,5 +1,14 @@
+enum class FragmentType{
+    LOGIN{ override fun get(): Fragment = LoginFragment() },
+    LOGINTYPE{ override fun get(): Fragment = LoginTypeFragment() },
+    MAIN{ override fun get(): Fragment = MainFragment() };
```
아하...
```
enum class FragmentType(val fragment :Fragment) {
    LOGIN(LoginFragment()), LOGINTYPE(LoginTypeFragment()), MAIN(MainFragment())
}
```
요건 몰랐군요..!

konempty replied:
enum도 변수를 가질수 있어요ㅋㅋㅋㅋ

konempty reviewed (changes requested):
조금만 더 수정 부탁드리고 소스코드의 루트 path가 java로 되어있는데 kotlin으로 수정하시면 좋지않을까 싶네요ㅋㅋㅋㅋㅋㅋ

konempty commented:
+ ktlint라는것에 대해서 알아보시고 프로젝트에 적용해보시면 좋을듯 하네요

minseonglove commented:
> * ktlint라는것에 대해서 알아보시고 프로젝트에 적용해보시면 좋을듯 하네요

웹에서도 eslint라고 문법 훈장님이 있었는데 코틀린도 있나보네요

konempty commented:
https://github.com/JLLeitschuh/ktlint-gradle 여기 참고해서 적용해보세요

konempty commented:
수정하시는 김에 앱 이름과 패키지 명도 수정을... 헤헤

minseonglove commented:
> 수정하시는 김에 앱 이름과 패키지 명도 수정을... 헤헤

헤헤헤

konempty commented:
앗 이건 따로 브랜치파서 해도 괜찮겠네요

minseonglove commented:
> 앗 이건 따로 브랜치파서 해도 괜찮겠네요

패키지명 변경은 깔끔하게 노 리뷰 머지로 갈까요

konempty commented:
기능변경은 아니니까 괜찮겠네요 조심해서 수정해주세요ㅋㅋㅋㅋ

minseonglove committed:
REFACTOR : konempty PR 리뷰 반영

> * ktlint에게 패배
> * FragmentType enum class 최적화
> * 소스 코드 root directory 이름을 java에서 kotlin으로 변경

konempty commented on app/build.gradle:
```diff
@@ -1,11 +1,15 @@
 plugins {
     id 'com.android.application'
     id 'org.jetbrains.kotlin.android'
+    id("org.jlleitschuh.gradle.ktlint") version "10.2.1"
```
코드 컨벤션을 맞추면 좋을것 같네요.ㅋㅋㅋㅋ

konempty reviewed (approved):
gradle파일에서 코드컨벤션이 조금씩 달라보이는게 보기 안좋아 보이긴 하지만 뭐 잘못된건 아니니 넘어가도록 하죠 ㅋㅋㅋㅋ 수고하셨습니다!

minseonglove committed:
REFACTOR : konempty PR 리뷰 반영

* ktlint에게 패배
* ktlint 플러그인을 프로젝트 gradle로 이동
